### DISCUSSION TITLE
Tracking class updates

### DISCUSSION THREAD

**Discussion by savchenko** [microsoft/pyright#8450] [Q&A]
Say we have a file `extensions.py` where:
```py
class Foo(some):
    def __init__(self, bar=None):
        self.bar = None

foo_instance = Foo()
```

in the `main.py` we:
```py
from extensions import foo_instance

foo_instance(bar=something)
```

Then, to avoid circular import, in `whatever.py` which is imported by `main.py`:
```
from extensions import foo_instance

foo_instance.something.some_method()
```

**The problem**

Pyright does not know that `foo_instance.something` exists and reports `reportOptionalMemberAccess`.

Is there a solution? Related to https://github.com/viniciuschiele/flask-apscheduler/issues/249

**Comment by erictraut** [comment]
In your example above, `foo_instance` isn't callable, so I don't see how this would work. I'm guessing that your example is incomplete.

If you want to use static typing, attributes on a class must be declared. If you're adding dynamic attributes to an object that are not declared, then a static type checker will not know about them and will complain if you attempt to access them. 

There are many ways to eliminate circular dependencies in your code, most of them involving refactoring. The technique you're using here looks very hacky. I'd need to understand your scenario better to offer more concrete suggestions.

**Reply by savchenko** [reply]
The example provided is a bastardised version of the [application factory](https://flask.palletsprojects.com/en/3.0.x/appcontext/), pattern widely used in Flask applications.

The issue I have linked in the original post is an official example of correct use by one of the Flask extensions.

I fully appreciate the challenge, as at time of import in `whatever.py`, linter would not know that `foo_instance.something` exists.

Here is what is happening:
![image](https://github.com/user-attachments/assets/9b0a33e8-287c-4f5c-b580-3d8de63711cc)

**Reply by erictraut** [reply]
This is still too abstract for me to offer much in the way of advice. My initial thought is that it looks like `Foo` should be a generic class.

This pattern also looks really fragile since `foo_instance` appears to be a singleton, and yet it's mutable. What if two modules import `foo_instance` and call` foo_instance(something)`? The last one wins? Perhaps it would make sense to revisit the entire design.

**Reply by savchenko** [reply]
Let me explain using concrete example. [Here](https://github.com/viniciuschiele/flask-apscheduler/blob/d5b6cdd10415a413b2773b7750d56b7dfed68018/flask_apscheduler/scheduler.py#L33-L49) is the `Foo` from above:

```python
class APScheduler(object):
    """Provides a scheduler integrated to Flask."""

    def __init__(self, scheduler=None, app=None):
        self.app = None
        if app:
            self.init_app(app)
```

and its [instance](https://github.com/viniciuschiele/flask-apscheduler/blob/d5b6cdd10415a413b2773b7750d56b7dfed68018/examples/application_factory/extensions.py#L3C1-L5C26):
```py
from flask_apscheduler import APScheduler

scheduler = APScheduler()
```

Now, let's look at init of an actual, working [application](https://github.com/viniciuschiele/flask-apscheduler/blob/d5b6cdd10415a413b2773b7750d56b7dfed68018/examples/application_factory/__init__.py#L21-L39):

```python
from .extensions import scheduler

def create_app():
    """Create a new app instance."""

    app = Flask(__name__)
    scheduler.init_app(app)

if __name__ == "__main__":
    app.run()
```

From the above we see that during the runtime `scheduler.app.*` will reference the current application.

However, from linter's perspective this is not the case and, for example, [this file](https://github.com/viniciuschiele/flask-apscheduler/blob/d5b6cdd10415a413b2773b7750d56b7dfed68018/examples/application_factory/events.py) lights up like a Christmas tree when checked by PyRight.

```py
from .extensions import scheduler

def job_missed(event):
    """Job missed event."""
    with scheduler.app.app_context():  # PyRight does not know about scheduler.app.*
        print(event)
```

**Reply by erictraut** [reply]
Yeah, this pattern won't work with static type checking. As I mentioned above, there are other reasons why this is a bad design pattern, so I encourage you to rethink it. If you can't change it for historical reasons, then you'll need to add some hacks if you want a static type checker to know about these types. This would likely involve calls to `typing.cast`.
